### ISSUE TITLE
twisted.web.client.readBody Deferred should be cancellable

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6686]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#6686|
|Type|enhancement|
|Created|2013-08-19 16:03:44Z|
|Branch|https://github.com/twisted/twisted/tree/readbody-cancellation-6686-3|

The `Deferred` returned from `readBody` should be cancellable, closing the connection to the HTTP server.

<details><summary>Searchable metadata</summary>

```
trac-id__6686 6686
type__enhancement enhancement
reporter__itamar itamar
priority__normal normal
milestone__ 
branch__branches_readbody_cancellation_6686_3 branches/readbody-cancellation-6686-3
branch_author__kaizhang__mithrandi kaizhang, mithrandi
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1376928224000000 1376928224000000
changetime__1480606528869170 1480606528869170
version__None None
owner__mithrandi mithrandi
cc__mithrandi cc__egraff
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|kaizhang set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|kaizhang commented|
|-|-|

(In [39902]) Branching to 'readbody-cancellation-6686'

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|kaizhang removed owner<br>kaizhang set status to `new`|
|-|-|

[Build Results](https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/readbody-cancellation-6686)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_5"></a>|@itamarst set owner to kaizhang|
|-|-|

1.The web client howto should be updated to mention cancellation support.
 1. Capitalize Deferred in the news file.
 1. The docstring for `readBody` should have the explanation of what cancellation does, so it shows up in API documentation.

Otherwise looks good. Fix the above, resubmit for review.

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|kaizhang set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)<a name="note_7"></a>|@mithrandi removed owner<br>@mithrandi set status to `new`|
|-|-|

Replying to [itamar](#note_5):
> Otherwise looks good. Fix the above, resubmit for review.

I believe I've fixed all of the issues raised.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph set owner to @mithrandi|
|-|-|

The fixes for the issues look good, but unfortunately this badly needs a merge forward.  Buildbots can't run it, and the docs are still in lore (!).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph commented|
|-|-|

Hit the 'submit' button a bit too fast there:

 * Thanks for updating this.
 * Please merge forward, force builds, and resubmit; I will likely sign off pretty quickly if I could look at a passing buildbot.

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)<a name="note_10"></a>|@mithrandi commented|
|-|-|

(In [43356]) Branching to readbody-cancellation-6686-2.

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)<a name="note_11"></a>|@mithrandi set owner to @glyph|
|-|-|

Okay, I merged forward. The buildbots seem to be happy modulo unrelated issues: http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/readbody-cancellation-6686-2

**Comment by twisted-trac** [comment]
|[<img alt="derwolfe's avatar" src="https://avatars.githubusercontent.com/u/641357?s=50" width="50" height="50">](https://github.com/derwolfe)<a name="note_12"></a>|@derwolfe set owner to @mithrandi|
|-|-|

Mithrandi - thanks for the fixes!

1. It looks like this branch just needs to be merged forward again against trunk to fix the test failures relating to inotify. 
2. The only issue I have found, is that I don't find the docstring for test_cancel very clear. I'd suggest rephrasing it to something like:
    "When L{Deferred.cancel} is called on a L{Deferred} returned by L{client.readBody}, the connection to the server will be aborted." 
But - if you think there is something that would be clearer; please use it. 

Once you've merged forward and changed the docstring, please resubmit for review.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_13"></a>|@glyph commented|
|-|-|

One minor comment about this - it actually makes `readBody`'s contract a bit stricter; previously, it would have worked with an object that provided only `ITransport`, but now, it requires one that implements `ITCPTransport`.

I don't remember off the top of my head what a canceler does if you raise an exception in it, but if it just results in the exception being logged then it's fine. If it actually delivers the error the application or ceases processing somewhere, then this might be a CompatibilityPolicy violation.

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)<a name="note_14"></a>|@mithrandi commented|
|-|-|

(In [43374]) Branching to readbody-cancellation-6686-3.

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)<a name="note_15"></a>|@mithrandi removed owner|
|-|-|

I merged forward, and forced builds again: http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/readbody-cancellation-6686-3 Still one bot failing, but still unrelated to this branch as far as I can tell.

I changed the code to handle a transport without `abortTransport` (as otherwise an exception would have been raised out of `deferred.cancel()`), but issue a deprecation warning when `readBody` is called with such a transport.

Unfortunately, if anyone actually gets that warning, it will probably not be clear as to what the problem is; since the transport is hidden behind the Agent and/or Response implementation being used, and not something the caller would directly be aware of.

I also made some minor grammatical changes to the docstring herrwolfe mentioned; I don't think it's unclear to use the verb "cancel" when referring to the method of the same name on `Deferred` (it's a new feature, so some people won't be familiar with it yet, but it's easy to search for), but the sentence construction wasn't quite right.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_16"></a>|@glyph commented|
|-|-|

That builder is failing with [#6027](https://github.com/twisted/twisted/issues/6027).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_17"></a>|@glyph set owner to @mithrandi|
|-|-|

Replying to [mithrandi](#note_15):
> I merged forward, and forced builds again: http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/readbody-cancellation-6686-3 Still one bot failing, but still unrelated to this branch as far as I can tell.

I re-forced them and it looks like they're all passing now.

> I changed the code to handle a transport without `abortTransport` (as otherwise an exception would have been raised out of `deferred.cancel()`), but issue a deprecation warning when `readBody` is called with such a transport.

Cool.

> Unfortunately, if anyone actually gets that warning, it will probably not be clear as to what the problem is; since the transport is hidden behind the Agent and/or Response implementation being used, and not something the caller would directly be aware of.

Honestly, I think it's pretty unlikely that many callers are likely to get this; you'd have to be using things in a fairly obscure way.  Given that you would have to go to some trouble to set up an alternate transport, you'd likely know what that was.  Users going through the more default channels will simply never see it.

If they are confused and get in touch with the mailing list because they don't understand the warning, great!  I just want to make sure we're staying in compliance with policy so that if people get an error when upgrading to 14.1+1, they can downgrade to 14.1 and see _some_ kind of warning.

> I also made some minor grammatical changes to the docstring herrwolfe mentioned; I don't think it's unclear to use the verb "cancel" when referring to the method of the same name on `Deferred` (it's a new feature, so some people won't be familiar with it yet, but it's easy to search for), but the sentence construction wasn't quite right.

It seems pretty clear to me.

I think it would be cool to land it now.  The only other enhancement I can think of would be to call loseConnection if that's available and abortConnection isn't, but that seems like maybe gold-plating :).

Please go ahead and land.

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)<a name="note_18"></a>|@mithrandi set status to `closed`|
|-|-|

(In [43388]) Merge readbody-cancellation-6686-3: Implement cancellation for readBody.

Author: kaizhang, mithrandi
Reviewers: itamar, glyph, herrwolfe
Fixes: [#6686](https://github.com/twisted/twisted/issues/6686)

**Comment by twisted-trac** [comment]
|[<img alt="egraff's avatar" src="https://avatars.githubusercontent.com/u/5569523?s=50" width="50" height="50">](https://github.com/egraff)<a name="note_19"></a>|@egraff commented|
|-|-|

This does not work as intended (see ticket:8227#comment:1).
